paasta_tools/deployd/queue.py

```diff
@@ -19,54 +29,54 @@ def __init__(self, client, path):
         self.local_state_condition = Condition()
         self.entry_nodes = []
         self.locked_entry_nodes = set()
-        self.update_local_state(None)
+        self._update_local_state(None)
 
-    def update_local_state(self, event):
+    def _update_local_state(self, event):
         with self.local_state_condition:
-            print("refreshing state")
-            if event is not None:
-                print(f"  event: {event}")
             entry_nodes = self.client.retry(
                 self.client.get_children,
                 self.entries_path,
-                watch=self.update_local_state,
+                watch=self._update_local_state,
             )
             self.entry_nodes = sorted(entry_nodes)
             self.locked_entry_nodes = set(
                 self.client.retry(
                     self.client.get_children,
                     self.locks_path,
-                    watch=self.update_local_state,
+                    watch=self._update_local_state,
                 )
             )
             self.local_state_condition.notify()
 
-    def qsize(self):
-        return len(self.entry_nodes)
+    def _format_timestamp(self, timestamp: float):
+        if not isinstance(timestamp, float):
+            raise TypeError("timestamp must be int or float")
+        if not (0 < timestamp < 9999999999.9995):
+            raise ValueError("timestamp must be between 0 and 9999999999.9995")
+
+        formatted = f"{timestamp:014.3f}"
+        assert len(formatted) == 14
 
-    @property
-    def queue(self):
-        return self.entry_nodes
+        return formatted
 
     # TODO: should priority just be a string?  can make it length 20 for two priorities?  or even make that configurable?
-    def put(self, value, priority):
-        self._check_put_arguments(value, priority)
+    def put(self, si: ServiceInstance):
+        bounce_by = self._format_timestamp(si.bounce_by)
+        wait_until = self._format_timestamp(si.wait_until)
+
         self.client.create(
-            f"{self.entries_path}/entry-{priority:010d}-", value=value, sequence=True
+            f"{self.entries_path}/entry-{bounce_by}-{wait_until}-",
+            value=json.dumps(si._asdict()).encode("utf-8"),
+            sequence=True,
         )
 
-    def _check_put_arguments(self, value, priority=100):
-        if not isinstance(value, bytes):
-            raise TypeError(f"value must be a byte string (got type {type(value)})")
-        elif not isinstance(priority, int):
-            raise TypeError(f"priority must be an int (got type {type(priority)})")
-        elif priority < 0 or priority > 9999999999:
-            raise ValueError(
-                f"priority must be between 0 and 9999999999 (got {priority})"
-            )
-
     @contextmanager
-    def get(self):  # TODO: timeout?
+    def get(self, block: bool = True, timeout: float = float("inf")):
+        self._update_local_state(None)
```

Comment:
I'm not sure this is necessary -- we have the watcher already, and this will mean that all deployd threads will end up blocking each other waiting for the condition lock

Reply:
Hm. I added this so that put() followed by an immediate nonblocking get() will always work, which is a situation that only really applies in testing. You're probably right about the extra blocking not being great; I guess I can just make the tests explicitly call _update_local_state and/or use small timeouts instead of true nonblocking.